torchdp/autograd_grad_sample.py

```diff
@@ -105,136 +115,160 @@ def _capture_activations(
     layer.activations = input[0].detach()
 
 
-def _capture_backprops(layer: nn.Module, _input, output):
-    """Append backprop to layer.backprops_list in backward pass."""
-    global _enforce_fresh_backprop
+def _capture_backprops(
+    layer: nn.Module,
+    _input: torch.Tensor,
+    output: torch.Tensor,
+    loss_type: str,
+    batch_dim: int,
+):
+    """Capture backprops in backward pass and store per-sample gradients."""
 
     if _hooks_disabled:
         return
 
-    if _enforce_fresh_backprop:
-        if hasattr(layer, "backprops_list"):
-            raise ValueError(
-                f"Seeing result of previous backprop, "
-                f"use clear_backprops(model) to clear"
-            )
-        _enforce_fresh_backprop = False
-
-    if not hasattr(layer, "backprops_list"):
-        layer.backprops_list = []
-    layer.backprops_list.append(output[0].detach())
+    backprops = output[0].detach()
+    _compute_grad_sample(layer, backprops, loss_type, batch_dim)
+    if hasattr(layer, "activations"):
+        del layer.activations
```

Comment:
nit: IMO, at one point in time we are going to have all activations, back-props, and grad_samples from one micro batch in memory, so really there is no point in deleting these unless we want to make space for other programs which we are not optimizing here for. let's just leave them be. deleting memory and recreating it only harms the perf.

Reply:
agreed